File: euler46.py
# Каково наименьшее нечетное составное число, которое нельзя записать
# в виде суммы простого числа и удвоенного квадрата?

from datetime import datetime

n = 10_000


def eratosfen(number):
    """Решето Эратосфена: генерация простых чисел вплоть до заданного числа"""
    sieve = list(range(number + 1))
    sieve[1] = 0
    for i in sieve:
        if i > 1:
            for j in range(2 * i, len(sieve), i):
                sieve[j] = 0
    return sieve

# 1 вариант
start_time = datetime.now()
goldbah_conjecture = set()
for p in set(eratosfen(n)):
    for k in range(int(n ** 0.5)):
        goldbah_conjecture.add(p + 2 * k * k)
else:
    print(min(set(range(3, n, 2)) - goldbah_conjecture))
print(datetime.now() - start_time)

# 2 вариант
start_time = datetime.now()
pass
print(datetime.now() - start_time)

File: go/euler46.go
// euler46 находит наименьшее нечетное составное число, которое нельзя
// записать как сумму простого числа и удвоенного квадрата.
//
// Решение задачи Эйлера №46.
package main

import (
	"fmt"
	"math"
	"time"
)

func eratosfen(number int) []bool {
	sieve := make([]bool, number+1)
	for i := range sieve {
		sieve[i] = true
	}
	sieve[0] = false
	sieve[1] = false
	for i := 2; i*i <= number; i++ {
		if sieve[i] {
			for j := i * i; j <= number; j += i {
				sieve[j] = false
			}
		}
	}
	return sieve
}

func euler46() {
	n := 10000

	startTime := time.Now()
	sieve := eratosfen(n)
	goldbahConjecture := make(map[int]bool)
	primes := make([]int, 0)
	for i := 2; i <= n; i++ {
		if sieve[i] {
			primes = append(primes, i)
			goldbahConjecture[i] = true
		}
	}
	for _, p := range primes {
		for k := 0; k*k*2 < n; k++ {
			goldbahConjecture[p+2*k*k] = true
		}
	}
	result := -1
	for i := 3; i < n; i += 2 {
		if !sieve[i] && !goldbahConjecture[i] {
			result = i
			break
		}
	}
	fmt.Println(result)
	fmt.Println(time.Since(startTime))

	startTime = time.Now()
	fmt.Println(time.Since(startTime))
}

func main() {
	euler46()
}
